### PR TITLE
Diagnostic tool to inspect storage node bandwidth allocation databases

### PR DESCRIPTION
**Pull request by aligeti**
https://storjlabs.atlassian.net/browse/V3-601

Diagnostic tool for DevOps

Added to the storagenode utility as a 'diag' command. To run at the command prompt run
> storagenode diag --> this runs a sql db query to read all the rows off of bandwidth_allocations table and populate them in a table. 

If you want to read another node, run below command 
> storagenode diag --base-path=/Users/xyz/.storj/capt/f37/data

It outputs in a table format.

### CODE REVIEW

File: cmd/storagenode/main.go [hunk 1]
```diff
@@ -88,6 +105,59 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	dbpath := filepath.Join(diagCfg.BasePath, "piecestore.db")
+	db, err := psdb.Open(context.Background(), "", dbpath)
+	if err != nil {
+		fmt.Println("Storagenode database couldnt open:", dbpath)
+		return err
+	}
+
+	//get all bandwidth aggrements entries
+	bwAgreements, err := db.GetBandwidthAllocations()
+	t := table.NewWriter()
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: Can't we just use tabwriter here? Or in other words, have you vetted the dependencies of pretty/table for security etc. and do we gain something significant from using it.

Comment (aligeti) [reply cmd/storagenode/main.go]: scanned thru the code in this package. on a quick scan, the packages imported are basic golang supported packages

Comment (egonelbre) [reply cmd/storagenode/main.go]: Can we just use tabwriter?

Comment (aligeti) [reply cmd/storagenode/main.go]: modified to use tab writer

File: go.mod [hunk 2]
```diff
@@ -105,6 +105,7 @@ require (
 	github.com/go-ole/go-ole v1.2.1 // indirect
 	github.com/hanwen/go-fuse v0.0.0-20181011180456-b760b55765be
 	github.com/inconshreveable/mousetrap v1.0.0 // indirect
+	github.com/jedib0t/go-pretty v4.0.0+incompatible
```
Comment (egonelbre) [inline go.mod]: This is not in use.

File: go.sum [hunk 3]
```diff
@@ -153,6 +153,8 @@ github.com/inconshreveable/mousetrap v1.0.0 h1:Z8tu5sraLXCXIcARxBp/8cbvlwVa7Z1NH
 github.com/inconshreveable/mousetrap v1.0.0/go.mod h1:PxqpIevigyE2G7u3NXJIT2ANytuPF1OarO4DADm73n8=
 github.com/jbenet/go-base58 v0.0.0-20150317085156-6237cf65f3a6 h1:4zOlv2my+vf98jT1nQt4bT/yKWUImevYPJ2H344CloE=
 github.com/jbenet/go-base58 v0.0.0-20150317085156-6237cf65f3a6/go.mod h1:r/8JmuR0qjuCiEhAolkfvdZgmPiHTnJaG0UXCSeR1Zo=
+github.com/jedib0t/go-pretty v4.0.0+incompatible h1:3I/BQ59W64OTQUCSwW2f7lX5xqBWjkqpr6XQ80tymS4=
+github.com/jedib0t/go-pretty v4.0.0+incompatible/go.mod h1:XemHduiw8R651AF9Pt4FwCTKeG3oo7hrHJAoznj9nag=
```
Comment (egonelbre) [inline go.sum]: This is not in use.

File: cmd/storagenode/main.go [hunk 4]
```diff
@@ -4,14 +4,21 @@
...
+	"text/tabwriter"
 
+	"github.com/golang/protobuf/proto"
 	"github.com/spf13/cobra"
 
 	"storj.io/storj/pkg/cfgstruct"
 	"storj.io/storj/pkg/kademlia"
+	"storj.io/storj/pkg/pb"
 	psserver "storj.io/storj/pkg/piecestore/rpc/server"
+	"storj.io/storj/pkg/piecestore/rpc/server/psdb"
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: This is stale package name.

File: cmd/storagenode/main.go [hunk 5]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+						} else {
+							satIDVal.GetActionCount++
+						}
+						satIDVal.TotalTransactions++
+						satelliteID[satIDKey] = satIDVal
+					}
+				}
+			}
+		}
+		fmt.Fprintln(w, currSatID, "\t", satelliteID[currSatID].TotalBytes, "\t", satelliteID[currSatID].TotalTransactions, "\t", satelliteID[currSatID].PutActionCount, "\t", satelliteID[currSatID].GetActionCount, "\t")
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: println adds spaces between statements. Use Fprint with `\n` at the end.

File: cmd/storagenode/main.go [hunk 6]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	const padding = 3
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, padding, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	fmt.Fprintln(w, "SatelliteID\tTotal\t# Of Transactions\tPUT Action\tGET Action\t")
+
+	// attributes per satelliteid
+	satelliteID := make(map[string]SatAttributes)
+	satAtt := SatAttributes{}
+	var currSatID, lastSatID string
+
+	for _, v := range bwAgreements {
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: Use a better variable name than `v`.

File: cmd/storagenode/main.go [hunk 7]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+		for _, j := range v {
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: Use a better variable name than `j`.

File: cmd/storagenode/main.go [hunk 8]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+				return err
+			}
+			total := rbad.GetTotal()
+
+			// deserializing pbad you get satelliteID, uplinkID, max size, exp, serial# & action
+			pbad := &pb.PayerBandwidthAllocation_Data{}
+			if err := proto.Unmarshal(rbad.GetPayerAllocation().GetData(), pbad); err != nil {
+				return err
+			}
+			currSatID = fmt.Sprintf("%s", pbad.GetSatelliteId())
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: Use `string(pbad.GetSatelliteId())`

File: cmd/storagenode/main.go [hunk 9]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	}
+
+	const padding = 3
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, padding, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	fmt.Fprintln(w, "SatelliteID\tTotal\t# Of Transactions\tPUT Action\tGET Action\t")
+
+	// attributes per satelliteid
+	satelliteID := make(map[string]SatAttributes)
+	satAtt := SatAttributes{}
+	var currSatID, lastSatID string
```
Comment (egonelbre) [inline cmd/storagenode/main.go]: These shouldn't be necessary, they significantly complicate the code.

File: cmd/storagenode/main.go (hunk 5 above)
Comment (egonelbre) [reply cmd/storagenode/main.go]: Also this breaks when the table is empty.

File: go.sum (hunk 3 above)
Comment (aligeti) [reply go.sum]: Done

File: cmd/storagenode/main.go (hunk 4 above)
Comment (aligeti) [reply cmd/storagenode/main.go]: Done

File: cmd/storagenode/main.go (hunk 6 above)
Comment (aligeti) [reply cmd/storagenode/main.go]: Renamed. rbaVal -> RentersBandwidthAgreement

File: cmd/storagenode/main.go (hunk 7 above)
Comment (aligeti) [reply cmd/storagenode/main.go]: Renamed. rbaDataVal -> RentersBandwidthAgreementDataValue

File: cmd/storagenode/main.go (hunk 8 above)
Comment (aligeti) [reply cmd/storagenode/main.go]: Done

File: cmd/storagenode/main.go (hunk 5 above)
Comment (aligeti) [reply cmd/storagenode/main.go]: modified as suggested

File: cmd/storagenode/main.go [hunk 10]
```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+
+			// deserializing pbad you get satelliteID, uplinkID, max size, exp, serial# & action
+			pbad := &pb.PayerBandwidthAllocation_Data{}
+			if err := proto.Unmarshal(rbad.GetPayerAllocation().GetData(), pbad); err != nil {
+				return err
+			}
+			currSatID = fmt.Sprintf("%s", pbad.GetSatelliteId())
+			action := pbad.GetAction()
+
+			if strings.Compare(currSatID, lastSatID) != 0 {
```
[the remark this answers is not in the document]
Comment (aligeti) [reply cmd/storagenode/main.go]: Done

File: go.sum (hunk 3 above)
Comment (egonelbre) [reply go.sum]: this is still here

File: go.mod (hunk 2 above)
Comment (egonelbre) [reply go.mod]: this is still here

Comment (aligeti) [reply go.mod]: Done.

File: go.sum (hunk 3 above)
Comment (aligeti) [reply go.sum]: Done

**Pull request merged**